feat: add time API guide (#3599)
JetBrains/kotlin-web-site: docs/topics/time.md

[//]: # (title: Time)

The Kotlin standard library gives you the tools you need to be able to calculate and measure time in different units.
Being able to measure time accurately is important for activities like:
  * Managing threads or processes
  * Collecting statistics
  * Detecting timeouts
  * Debugging

By default, time is measured using a monotonic time source but other time sources can be configured.

## Durations

To represent an amount of time, the standard library has the [`Duration`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/)
class. A `Duration` can be expressed in the following units from the [`DurationUnit`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration-unit/)

> Enum class:

Reviewer: enum class

  * `NANOSECONDS`
  * `MICROSECONDS`
  * `MILLISECONDS`
  * `SECONDS`
  * `MINUTES`
  * `HOURS`
  * `DAYS`

`Durations` can also be zero, negative or infinite.

### Calculate duration

> To create an instance of the `Duration` class, use the [extension properties](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/#companion-object-properties) 

Reviewer: Duration is a value class and therefore it doesn't have an instance identity. Nothing is created (in sense of object creation) in the result of these calls.

Author: I updated the text. Is it OK now?

available for `Int`, `Long`, and `Double` types: `nanoseconds`, `microseconds`, `milliseconds`, `seconds`, `minutes`, 

> `hours`, and `days`.

Reviewer: May be important to reiterate that days are always 24 hour periods and not calendar days

For example:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val fiveHundredMilliseconds: Duration = 500.milliseconds
    val fiveSeconds: Duration = 5.seconds
    val tenMinutes: Duration = 10.minutes

    println(fiveHundredMilliseconds) // 500ms
    println(fiveSeconds)             // 5s
    println(tenMinutes)              // 10m
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> To create a `Duration` of zero seconds, use the `ZERO` property:

Reviewer: `ZERO` may be a convenient constant, but I wouldn't give a strong recommendation to use it certainly if you need a duration of zero seconds.

Author: OK, I'll remove this section.

```kotlin
import kotlin.time.*

> import kotlin.time.Duration.Companion.ZERO

Reviewer: Better not to import it, and use just Duration.ZERO

fun main() {
    //sampleStart
    val zero: Duration = ZERO

    println(zero) // 0s
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

You can also perform basic arithmetic with `Durations`:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.seconds

fun main() {
    //sampleStart
    val fiveSeconds: Duration = 5.seconds
    val thirtySeconds: Duration = 30.seconds

    println(fiveSeconds + thirtySeconds)
    // 35s
    println(thirtySeconds -  fiveSeconds)
    // 25s
    println(fiveSeconds * 2)
    // 10s
    println(thirtySeconds / 2)
    // 15s
    println(thirtySeconds / fiveSeconds)
    // 6.0
    println(-thirtySeconds)
    // -30s
    println((-thirtySeconds).absoluteValue)
    // 30s
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> ### Print duration

Reviewer: I would center this around getting a string representation of a duration and printing it may be one of the use cases of it. Other use cases may be: serialization and transferring, storage.

Author: I updated this section as you suggested. Is it OK now?

To print a `Duration`, you have already seen that you can use `println()` and `print()` functions. By default, the

> time is reported using each unit that is present. For example: `1h 0m 45.677s`.

Reviewer: May be also useful to show how negative durations with multiple components are formatted (with minus and parentheses)

To configure the output, use the `.toString()` function with your desired `Duration.Unit` and number of decimal places
as function parameters:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.DurationUnit

fun main() {
    //sampleStart
    val fiveSeconds: Duration = 5.seconds
    
    // Print in minutes with 2 decimal places
    println(fiveSeconds.toString(DurationUnit.MINUTES, 2))

>     // 0,08m

Reviewer: Better to use a duration with some sub-second part and format it to seconds for example:
```
5887.milliseconds.toString(DurationUnit.SECONDS, 2) // 5.89s
```

    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

To print the `Duration` as an [ISO-8601-compatible](https://en.wikipedia.org/wiki/ISO_8601) string, use the [`toIsoString()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/to-iso-string.html)
function:

```kotlin
import kotlin.time.Duration.Companion.seconds

fun main() {
    //sampleStart
    println(86420.seconds.toIsoString()) // PT24H0M20S
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Convert duration

> To express your `Duration` in a different `Duration.Unit` you can use the properties:

Reviewer: Not sure if "express" is correct here.

Author: Changed to "convert".

* `inWholeNanoseconds`
* `inWholeMicroseconds`
* `inWholeSeconds`
* `inWholeMinutes`
* `inWholeHours`
* `inWholeDays`

For example:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val thirtyMinutes: Duration = 30.minutes
    println(thirtyMinutes.inWholeSeconds)
    // 1800
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

Alternatively, you can use your desired `Duration.Unit` as a function parameter in the following extension functions:
* `toInt()`
* `toDouble()`
* `toLong()`

For example:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.DurationUnit

fun main() {
    //sampleStart
    val thirtyMinutes: Duration = 30.minutes

>     println(thirtyMinutes.toDouble(DurationUnit.SECONDS))

Reviewer: For converting to double better to use a value that actually produces a fractional part.

    //1800.0
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Compare duration

To check if `Durations` are equal, use the equality operator (`==`):

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val thirtyMinutes: Duration = 30.minutes
    val halfHour: Duration = 0.5.hours
    println(thirtyMinutes == halfHour)
    // true
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

To compare `Durations`, use the comparison operators (`<`, `>`):

```kotlin
import kotlin.time.Duration.Companion.microseconds
import kotlin.time.Duration.Companion.nanoseconds

fun main() {
    //sampleStart
    println(3000.microseconds < 25000.nanoseconds)
    // false
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Break duration into components

> To break down a `Duration` into its time components and perform a further action, use the `toComponents()` function.

Reviewer: "use one of overloads of `toComponents` function." This requires an explanation how these overloads differ from each other.

[//]: # (title: Time)

The Kotlin standard library gives you the tools you need to be able to calculate and measure time in different units.
Being able to measure time accurately is important for activities like:
  * Managing threads or processes
  * Collecting statistics
  * Detecting timeouts
  * Debugging

By default, time is measured using a monotonic time source but other time sources can be configured.

## Durations

To represent an amount of time, the standard library has the [`Duration`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/)
class. A `Duration` can be expressed in the following units from the [`DurationUnit`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration-unit/)
enum class:
  * `NANOSECONDS`
  * `MICROSECONDS`
  * `MILLISECONDS`
  * `SECONDS`
  * `MINUTES`
  * `HOURS`
  * `DAYS`

> `Durations` can also be zero, negative or infinite.

Reviewer: Not quite clear what "also" means.

Author: It's meant to indicate additional information about the characteristics of the `Durations` class. I've rephrased it to:
> In addition, `Durations` can be zero, negative or infinite.

Is this more clear for you?

available for `Int`, `Long`, and `Double` types: `nanoseconds`, `microseconds`, `milliseconds`, `seconds`, `minutes`, 
`hours`, and `days`.

> Days refer to periods of 24 hours. They are not calendar days.
> 
{type = "tip"}

For example:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val fiveHundredMilliseconds: Duration = 500.milliseconds
    val fiveSeconds: Duration = 5.seconds
    val tenMinutes: Duration = 10.minutes

    println(fiveHundredMilliseconds) // 500ms
    println(fiveSeconds)             // 5s
    println(tenMinutes)              // 10m
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

You can also perform basic arithmetic with `Durations`:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.seconds

fun main() {
    //sampleStart
    val fiveSeconds: Duration = 5.seconds
    val thirtySeconds: Duration = 30.seconds

    println(fiveSeconds + thirtySeconds)
    // 35s

>     println(thirtySeconds -  fiveSeconds)

Reviewer: extra space

    // 25s
    println(fiveSeconds * 2)
    // 10s
    println(thirtySeconds / 2)
    // 15s
    println(thirtySeconds / fiveSeconds)
    // 6.0
    println(-thirtySeconds)
    // -30s
    println((-thirtySeconds).absoluteValue)
    // 30s
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Get string representation

It can be useful to have a string representation of a `Duration` so that you can print, serialize, transfer, or store it.

To get a string representation, use the `.toString()` function. By default, the time is reported using each unit that is
present. For example: `1h 0m 45.677s` or `-(6d 5h 5m 28.284s)`

To configure the output, use the `.toString()` function with your desired `Duration.Unit` and number of decimal places
as function parameters:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.DurationUnit

fun main() {
    //sampleStart
    // Print in seconds with 2 decimal places
    println(5887.milliseconds.toString(DurationUnit.SECONDS, 2))
    // 5.89s
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> To print the `Duration` as an [ISO-8601-compatible](https://en.wikipedia.org/wiki/ISO_8601) string, use the [`toIsoString()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/to-iso-string.html)

Reviewer: Here again printing is only one of use cases of getting a string representations.

function:

```kotlin
import kotlin.time.Duration.Companion.seconds

fun main() {
    //sampleStart
    println(86420.seconds.toIsoString()) // PT24H0M20S
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Convert duration

> To convert your `Duration` into a different `Duration.Unit`, you can use the properties:

Reviewer: There's no such type as `Duration.Unit`. Perhaps `DurationUnit`?

[//]: # (title: Time)

> The Kotlin standard library gives you the tools you need to be able to calculate and measure time in different units.

Reviewer 2: ```suggestion
The Kotlin standard library gives you the tools to calculate and measure time in different units.
```

> Being able to measure time accurately is important for activities like:

Reviewer 2: ```suggestion
Accurately time measurement is important for:
```

  * Managing threads or processes
  * Collecting statistics
  * Detecting timeouts
  * Debugging

> By default, time is measured using a monotonic time source but other time sources can be configured.

Reviewer 2: ```suggestion
By default, time is measured using a monotonic time source, but other time sources can be configured.
```

For more information, see [Create time source](#create-time-source).

## Durations

To represent an amount of time, the standard library has the [`Duration`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/)
class. A `Duration` can be expressed in the following units from the [`DurationUnit`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration-unit/)
enum class:
  * `NANOSECONDS`
  * `MICROSECONDS`
  * `MILLISECONDS`
  * `SECONDS`
  * `MINUTES`
  * `HOURS`
  * `DAYS`

> `Durations` can be positive, negative, zero, positive infinity, or negative infinity.

Reviewer 2: ```suggestion
A `Duration` can be positive, negative, zero, positive infinity, or negative infinity.
```

`hours`, and `days`.

> Days refer to periods of 24 hours. They are not calendar days.
> 
{type = "tip"}

For example:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.nanoseconds
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.days

fun main() {
    //sampleStart
    val fiveHundredMilliseconds: Duration = 500.milliseconds
    val zeroSeconds: Duration = 0.seconds
    val tenMinutes: Duration = 10.minutes
    val negativeNanosecond: Duration = (-1).nanoseconds
    val infiniteDays: Duration = Double.POSITIVE_INFINITY.days
    val negativeInfiniteDays: Duration = Double.NEGATIVE_INFINITY.days

    println(fiveHundredMilliseconds) // 500ms
    println(zeroSeconds)             // 0s
    println(tenMinutes)              // 10m
    println(negativeNanosecond)      // -1ns
    println(infiniteDays)            // Infinity
    println(negativeInfiniteDays)    // -Infinity
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-create-duration"}

> You can also perform basic arithmetic with `Durations`:

Author: Because the formatting looks a bit ugly as `Duration`s, I decided to change it to `Duration` objects instead. I've applied this change for all the other similar cases of this 😄

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.seconds

fun main() {
    //sampleStart
    val fiveSeconds: Duration = 5.seconds
    val thirtySeconds: Duration = 30.seconds

    println(fiveSeconds + thirtySeconds)
    // 35s
    println(thirtySeconds - fiveSeconds)
    // 25s
    println(fiveSeconds * 2)
    // 10s
    println(thirtySeconds / 2)
    // 15s
    println(thirtySeconds / fiveSeconds)
    // 6.0
    println(-thirtySeconds)
    // -30s
    println((-thirtySeconds).absoluteValue)
    // 30s
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-create-duration-arithmetic"}

### Get string representation

It can be useful to have a string representation of a `Duration` so that you can print, serialize, transfer, or store it.

To get a string representation, use the `.toString()` function. By default, the time is reported using each unit that is
present. For example: `1h 0m 45.677s` or `-(6d 5h 5m 28.284s)`

> To configure the output, use the `.toString()` function with your desired `DurationUnit` and number of decimal places

Reviewer 2: ```suggestion
To configure the output, use the `.toString()` function with your desired `DurationUnit` and a number of decimal places
```

Author: In this case, we don't need an article here. It's more natural without it.

}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-iso-string-representation"}

### Convert duration

To convert your `Duration` into a different `DurationUnit`, you can use the properties:
* `inWholeNanoseconds`
* `inWholeMicroseconds`
* `inWholeSeconds`
* `inWholeMinutes`
* `inWholeHours`
* `inWholeDays`

For example:

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val thirtyMinutes: Duration = 30.minutes
    println(thirtyMinutes.inWholeSeconds)
    // 1800
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-convert-duration"}

Alternatively, you can use your desired `DurationUnit` as a function parameter in the following extension functions:
* `.toInt()`
* `.toDouble()`
* `.toLong()`

For example:

```kotlin
import kotlin.time.Duration.Companion.seconds
import kotlin.time.DurationUnit

fun main() {
    //sampleStart
    println(270.seconds.toDouble(DurationUnit.MINUTES))
    // 4.5
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-convert-duration-extension"}

### Compare duration

> To check if `Durations` are equal, use the equality operator (`==`):

Reviewer 2: ```suggestion
To check if `Duration`s are equal, use the equality operator (`==`):
```

```kotlin
import kotlin.time.Duration
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes

fun main() {
    //sampleStart
    val thirtyMinutes: Duration = 30.minutes
    val halfHour: Duration = 0.5.hours
    println(thirtyMinutes == halfHour)
    // true
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-equality-duration"}

> To compare `Durations`, use the comparison operators (`<`, `>`):

Reviewer 2: ```suggestion
To compare `Duration`s, use the comparison operators (`<`, `>`):
```

For example:

```kotlin
import kotlin.time.measureTimedValue

fun main() {
    //sampleStart
    val (value, timeTaken) = measureTimedValue {
        Thread.sleep(100)
        42
    }
    println(value)     // 42
    println(timeTaken) // e.g. 103 ms
    //sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-measure-timed-value"}

By default, both functions use a monotonic time source.

### Mark moments in time

To mark a specific moment in time, use the [`TimeSource`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-source/)
interface and the [`markNow()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-source/mark-now.html) function
to create a [`TimeMark`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-mark/):

```kotlin
import kotlin.time.*
fun main() {
//sampleStart
   val timeSource = TimeSource.Monotonic
   val mark = timeSource.markNow()
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-mark-moment"}

### Measure differences in time

> To measure differences between `TimeMarks` from the same time source, use the subtraction operator (`-`).

Reviewer 2: ```suggestion
To measure differences between `TimeMark`s from the same time source, use the subtraction operator (`-`).
```

> To compare `TimeMarks` from the same time source, use the comparison operators (`<`, `>`).

Reviewer 2: ```suggestion
To compare `TimeMark`s from the same time source, use the comparison operators (`<`, `>`).
```

For example:

```kotlin

> import kotlin.time.*

Reviewer 2: ```suggestion
import kotlin.time.*

```

To check if a deadline has passed or a timeout has been reached, use the [`hasPassedNow()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-mark/has-passed-now.html)
and [`hasNotPassedNow()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-mark/has-not-passed-now.html) 
extension functions:

```kotlin
import kotlin.time.*
import kotlin.time.Duration.Companion.seconds

fun main() {
//sampleStart
   val timeSource = TimeSource.Monotonic
   val mark1 = timeSource.markNow()
   val fiveSeconds: Duration = 5.seconds
   val mark2 = mark1 + fiveSeconds

   // It hasn't been 5 seconds yet
   println(mark2.hasPassedNow())
   // false
  
   // Wait six seconds
   Thread.sleep(6000)
   println(mark2.hasPassedNow())
   // true

//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-deadline=passed"}

## Time sources

By default, time is measured using a monotonic time source. Monotonic time sources only move forward and are not affected
by variations like timezones. An alternative to monotonic time is elapsed real time which is also known as wall-clock time.

> Elapsed real time is time measured relative to another point in time.

Reviewer 2: ```suggestion
Elapsed real time is measured relative to another point in time.
```

> This section contains a list of default monotonic time sources per platform, as well as instructions on how to [create

Reviewer 2: I don't think this paragraph is necessary, the section is not that long to introduce the list of contents

your own time source](#create-time-source).

### Default time sources per platform

This table explains the default source of monotonic time for each platform:

| Platform      | Source |
|---------------|---|
| Kotlin/JVM    | `System.nanoTime()`|
| Kotlin/JS (Node.js) | `process.hrtime()`|

> | Kotlin/JS (Browser) | `window.performance.now()` or `Date.now()`|

Reviewer 2: ```suggestion
| Kotlin/JS (browser) | `window.performance.now()` or `Date.now()`|
```

   println(mark2.hasPassedNow())
   // false
  
   // Wait six seconds
   Thread.sleep(6000)
   println(mark2.hasPassedNow())
   // true

//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-time-deadline=passed"}

## Time sources

By default, time is measured using a monotonic time source. Monotonic time sources only move forward and are not affected
by variations like timezones. An alternative to monotonic time is elapsed real time which is also known as wall-clock time.
Elapsed real time is time measured relative to another point in time.

This section contains a list of default monotonic time sources per platform, as well as instructions on how to [create
a non-monotonic time source](#create-non-monotonic-time-source).

### Monotonic time sources per platform

This table explains the default source of monotonic time for each platform:

| Platform      | Source |
|---------------|---|
| Kotlin/JVM    | `System.nanoTime()`|
| Kotlin/JS (Node.js) | `process.hrtime()`|
| Kotlin/JS (Browser) | `window.performance.now()` or `Date.now()`|
| Kotlin/Native | `std::chrono::high_resolution_clock` or `std::chrono::steady_clock`|

> ### Create non-monotonic time source

Reviewer: This is a **monotonic** time source, as it can be seen from its name: `RealtimeMonotonicTimeSource`

There are some cases where you might want to use a non-monotonic time source. For example in Android, `System.nanoTime()`
only counts time while the device is active. It loses track of time when the device enters deep sleep. To keep track of
time while the device is in deep sleep, you can create a time source that uses [`SystemClock.elapsedRealtimeNanos()`](https://developer.android.com/reference/android/os/SystemClock#elapsedRealtimeNanos()):

```kotlin
object RealtimeMonotonicTimeSource : AbstractLongTimeSource(DurationUnit.NANOSECONDS) {
    override fun read(): Long = SystemClock.elapsedRealtimeNanos()
}
```

> Then you can declare your non-monotonic time source when you want to make time measurements:

Reviewer: Not "declare", but "use"

```kotlin
import kotlin.time.measureTime

//sampleStart
object RealtimeMonotonicTimeSource : AbstractLongTimeSource(DurationUnit.NANOSECONDS) {
    override fun read(): Long = SystemClock.elapsedRealtimeNanos()

> }

Reviewer: Since the sample is not runnable, it doesn't make sense to repeat the declaration of `RealtimeMonotonicTimeSource`.
